abi.md

```diff
@@ -69,7 +77,7 @@ packed vector of equivalent size.  For example, consider the following FIRRTL:
 
 ```FIRRTL
 circuit Top :
-  module Top :
+  public module Top :
     output out: UInt<16>
     input b: UInt<32>
 ```
```

Comment:
Is it intended non-main modules should be something a separate FIRRTL file can make use of via, say, `extmodule`?

Since there's no ability to specify the circuit name for an extmodule (the circuit the module is "imported" from?) ABI relying on circuit name across boundaries are unable to match up.  In particular I mean the ref ABI in the text below.  This is an issue an practice and would be good to fix as part of this.

Since we're requiring the public module name to be the verilog name , and only really "link" at Verilog level, it seems any namespacing provided by including the circuit name is not useful in practice as there would be a naming conflict of Verilog modules already (although I do like the idea, esp leaving door open for such).

While this ABI is clear about a single unit ("what is required/can be expected re:Verilog output"), it doesn't really say how a FIRRTL compiler is supposed to make use of this on the instantiation-side from a separate FIRRTL (or anything else wishing to make use of a FIRRTL-ABI-built "circuit"), so maybe there's some room here-- in addition to how they're used ("module name" needs to be "defname" in practice) it's also unspecified how they're gathered/included such that the macros are available reliably (left to build flow)...  (not unreasonably so, bear with me please)...

One could imagine requiring telling the build flow/driver/linker+compiler that module X is from circuit Y or something (since build flow knows what circuits it built anyway, even if individual builds either don't have that information or we'd like to be able to change the definitions being used (so don't want to bake circuit name in the code)?), leaving circuit name in the filenames for co-located multiple implementations of a thing but macros are just `ref_<module>_` as within any single "linked" design the module name must be unique.

Similar but maybe simpler possibility is to just drop the circuit portion of the ref ABI (`ref_<module>` prefix for files/macros), or if wishing backwards-compat with things built using ABI in current document, then treat all modules as being within their own circuit `ref_<module>_<module>`.

I'll think some more, maybe we should discuss a bit.  Thanks.

Reply:
This makes a lot of sense. Thanks for writing this up. It took me a couple of days to process. 😅 😓 

There's two things here:

1. To properly handle linking across a boundary, only the _module name_ should be used. This implies that some `extmodule Foo` should match up to some `public module Foo` in another compilation or it's Verilog/VHDL equivalent. This _shouldn't_ specify the circuit name as the circuit where this comes from is irrelevant. It's up to the user to figure this out and link it properly.
2. There is a namespacing problem with private modules. At minimum, these should have the circuit name included. At best, these should use some name mangling which cannot be linked from another FIRRTL compilation. E.g., Verilog allows for `$` in module names, but the FIRRTL spec disallows this. We could use this fact to generate a unique private module name which cannot conflict with something else like `<circuit>$<module>`. It's not pretty, but would work. The complication here is that if the spec _isn't_ silent about this, then it imposes an implicit private module ABI which could be used to link things silently (except that the private ABI can't be accessed from FIRRTL due to the use of `$`?).

WDYT?

Reply:
The only problem with using the bare module name without the circuit for this is that conflicts are high. clang/gcc avoid this on the C++ side with more complicated name mangling. We could do something like that that includes the names of the ports and their types in a mangled way. This would be super ugly... Perhaps its best to use the module name, but expect/encourage people to use names that are prefixed intelligently and/or use very descriptive names.

Reply:
I added language that makes things more "public module-centric" in https://github.com/chipsalliance/firrtl-spec/pull/130/commits/ce8aa58bf765c673bb26bcb89b764bd560d1c2e3.  Primarily, this means that the circuit name has no real effect anymore. The only place I see that being used in in the implementation-defined convention for private modules.

E.g., consider the following circuit:

```
circuit Foo:
  module Baz:

  public module Bar:
    inst baz of Baz
```

This must produce a Verilog module called `Bar` in a file `Bar.sv`. If it produces a Verilog module for FIRRTL module `Baz`, it should mangle its name to avoid a collision with another Verilog module. Some examples/ideas:

- `Foo_Baz` (simple, to the point)
- `_Foo_Baz`
- `Foo$Baz` (has the benefit of being legal in FIRRTL, but not in Verilog)

The primary benefit of using the circuit name here is that it is trivial to change without an effect on the rest of the design now. This then fills one of, though not all of, the rolls of the `NestedPrefixModulesAnnotation`.

Reply:
This sounds great!! I don't know how to "manage" the versioning details here -- I'd kinda thought the ABI was mostly versioned orthogonal to the FIRRTL spec, so this change should be ABI v3 and v4 or something?

Whatever seems best, I'm not sure how important that is presently, or if I'm missing something.

Anyway I think everything you've changed and your solutions/answers look great and excited for this direction.  Thanks! :rocket:

Reply:
Oh, looks like ABI is versioned with FIRRTL spec ? And v1/v2 are parts of that or something.  So probably nevermind :+1:.

Reply:
I'm glad this makes sense. Thanks for checking the details.

Yes, the ABI _document_ is co-versioned with the FIRRTL spec. However, the actual ABI can have different "versions" within it. ABI v1/v2 is really "lowered aggregates" and "preserved aggregates". This gets a little complicated as maybe we should then do the same thing for the ref and group ABIs given that this patch makes a change to them. I think we can dodge this and change it with the co-versioned version.

I.e., if you are compiling FIRRTL less than version 3 then you should be generating the files with a certain format. If you are compiling FIRRTL version 4 or greater, then you should generate the new format. I would like to avoid this complexity in CIRCT and do a hard switch to version 4. If we have to, we can add in support for version 3 if there is a use case. However, I'm not planning to add it.